store/sqlstore/group_store.go

```diff
@@ -116,6 +116,136 @@ func (s *SqlGroupStore) Create(group *model.Group) (*model.Group, error) {
 	return group, nil
 }
 
+func (s *SqlGroupStore) CreateWithUserIds(g *model.GroupWithUserIds) (*model.Group, error) {
+	if g.Id != "" {
+		return nil, store.NewErrInvalidInput("Group", "id", g.Id)
+	}
+
+	// Check if group values are formatted correctly
+	if err := g.IsValidForCreate(); err != nil {
+		return nil, err
+	}
+
+	// Check Users exist
+	if err := s.checkUsersExist(g.UserIds); err != nil {
+		return nil, err
+	}
+
+	g.Id = model.NewId()
+	g.CreateAt = model.GetMillis()
+	g.UpdateAt = g.CreateAt
+
+	groupInsertQuery, groupInsertArgs, err := s.getQueryBuilder().
+		Insert("UserGroups").
+		Columns("Id", "Name", "DisplayName", "Description", "Source", "RemoteId", "CreateAt", "UpdateAt", "DeleteAt", "AllowReference").
+		Values(g.Id, g.Name, g.DisplayName, g.Description, g.Source, g.RemoteId, g.CreateAt, g.UpdateAt, 0, g.AllowReference).
+		ToSql()
+	if err != nil {
+		return nil, err
+	}
+
+	usersInsertQuery, usersInsertArgs, err := s.buildInsertGroupUsersQuery(g.Id, g.UserIds)
+	if err != nil {
+		return nil, err
+	}
+
+	txn, err := s.GetMasterX().Beginx()
+	if err != nil {
+		return nil, err
+	}
+	defer finalizeTransactionX(txn)
+	// Create a new usergroup
+	if _, err = txn.Exec(groupInsertQuery, groupInsertArgs...); err != nil {
+		if IsUniqueConstraintError(err, []string{"Name", "groups_name_key"}) {
+			return nil, store.NewErrUniqueConstraint("Name")
+		}
+		return nil, errors.Wrap(err, "failed to save Group")
+	}
+	// Insert the Group Members
+	if _, err = executePossiblyEmptyQuery(txn, usersInsertQuery, usersInsertArgs...); err != nil {
+		return nil, err
+	}
+
+	// Get the new Group along with the member count
+	groupGroupQuery := `
+		SELECT
+			UserGroups.*,
+			A.Count AS MemberCount
+		FROM
+			UserGroups
+			INNER JOIN (
+				SELECT
+					UserGroups.Id,
+					COUNT(GroupMembers.UserId) AS Count
+				FROM
+					UserGroups
+					LEFT JOIN GroupMembers ON UserGroups.Id = GroupMembers.GroupId
+				WHERE
+					UserGroups.Id = ?
+				GROUP BY
+					UserGroups.Id
+				ORDER BY
+					UserGroups.DisplayName,
+					UserGroups.Id
+				LIMIT
+					? OFFSET ?
+			) AS A ON UserGroups.Id = A.Id
+		ORDER BY
+			UserGroups.CreateAt DESC`
+	var newGroup group
+	if err = txn.Get(&newGroup, groupGroupQuery, g.Id, 1, 0); err != nil {
+		return nil, err
+	}
+	if err = txn.Commit(); err != nil {
+		return nil, err
+	}
+	return newGroup.ToModel(), nil
+}
+
+func (s *SqlGroupStore) checkUsersExist(userIDs []string) error {
```

Comment:
I didn't see a unit test for this. Looks like a great candidate.

Reply:
Yeah. The behaviour is tested at https://github.com/mattermost/mattermost-server/pull/18839/files#diff-296cb9fe9469829f3e083acdb746280d12e303fada0491fa1225710004e5a780R345-R359. The one hesitation I have about exporting this to unit test it would be that it doesn't really belongs in the group store. It's more of a user store thing. Other options would be:
1. verify the user existence at another tier, [like the App](https://github.com/mattermost/mattermost-server/blob/custom_groups/app/group.go#L106)
2. add a foreign key constraint to group members to get the integrity check for free

I actually quite like option 2 because it's more resilient, more far-reaching, and avoids extra DB queries. That's something I think would be a good separate initiative. Thoughts?